### PR TITLE
Run check-config in a separate presubmit test job

### LINKED ISSUE #1686: Presubmit tests in test-infra should also include testing for staging prow configuration
Since staging prow configuration file is also generated from the config generator tool, if there are some changes to the tool, the auto-generated staging prow configuration will also be changed.

Currently in presubmit tests we are only checking the prod prow configuration file, see https://github.com/knative/test-infra/blob/64615f92c4ebdd4e4423015ffc0db45877660a0d/test/presubmit-tests.sh#L39, we should also check staging prow configuration file here.

/kind good-first-issue

### PR DESCRIPTION
**Pull request by chizhg**
**What this PR does, why we need it**:
Run check-config in a separate presubmit test job, other than build tests. This allows people to find out the error reason more clearly.
Also fix #1686 to also check prow-staging config files.

### CODE REVIEW

File: config/prow_common.mk [hunk 1]
```diff
@@ -153,47 +153,11 @@ update-prow-cluster: update-almost-all-cluster-deployments update-all-boskos-dep
 # In emergency, could easily edit this file, deleting all these lines
 confirm-master:
 	@if git diff-index --quiet HEAD; then true; else echo "Git working space is dirty -- will not update server"; false; fi;
-ifneq ("$(shell git branch --show-current)","master")
```
Comment (coryrc) [inline config/prow_common.mk]: `git rev-parse --abbrev-ref HEAD` can be affected by a dirty working space or something
`git branch --show-current` is exactly designed for this purpose

File: config/prow/Makefile [hunk 2]
```diff
@@ -15,10 +15,8 @@
 PROJECT       := knative-tests
 PROW_GCS      := knative-prow
 PROW_HOST     := https://prow.knative.dev
-TESTGRID_GCS  := knative-testgrid
```
Comment (coryrc) [inline config/prow/Makefile]: Could you instead leave this here and remove the TESTGRID_GCS from prow_common.mk? It's what I meant to do but overlooked.

Removing TESTGRID_CONFIG is good though, you should continue removing it.

File: config/prow/jobs/config.yaml [hunk 3]
```diff
@@ -3428,6 +3428,70 @@ presubmits:
...
+  - name: pull-knative-test-infra-check-config
+    agent: kubernetes
+    context: pull-knative-test-infra-check-config
+    always_run: true
+    rerun_command: "/test pull-knative-test-infra-check-config"
+    trigger: "(?m)^/test (all|pull-knative-test-infra-check-config),?(\\s+|$)"
+    decorate: true
+    path_alias: knative.dev/test-infra
+    branches:
+    - "release-0.10"
```
Comment (coryrc) [inline config/prow/jobs/config.yaml]: This is weird..... we don't have branches in test-infra

File: test/presubmit-config-check.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+source $(dirname $0)/../scripts/library.sh
+
+set -e
+
+trap 'echo "Please rerun \`make -C config/prow config\`."' ERR
+header "Checking generated config for production prow and testgrid"
+export PROW_CONFIG=$(mktemp)
+export PROW_JOB_CONFIG=$(mktemp)
+export PROW_PLUGINS=$(mktemp)
+export TESTGRID_CONFIG=$(mktemp)
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: You aren't actually writing to these files. This bash script has to invert the logic of the Makefile; first copy each of the files to the temp file, run make config, then compare them.

File: test/presubmit-config-check.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+diff --ignore-matching-lines="^# Copyright " "${TESTGRID_CONFIG}" "config/prow/testgrid/testgrid.yaml"
+
+trap 'echo "Please rerun \`make -C config/prow-staging config\`."' ERR
+header "Checking generated config for staging prow"
+subheader "Regenerating config for staging prow"
+make -C config/prow-staging config
+subheader "Comparing the generated config files with the existing config files"
+diff --ignore-matching-lines="^# Copyright " "${PROW_CONFIG}" "config/prow-staging/core/config.yaml"
+diff --ignore-matching-lines="^# Copyright " "${PROW_JOB_CONFIG}" "config/prow-staging/jobs/config.yaml"
+diff --ignore-matching-lines="^# Copyright " "${PROW_PLUGINS}" "config/prow-staging/core/plugins.yaml"
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: This should be able to be function-ized

File: test/presubmit-tests.sh [hunk 6]
```diff
@@ -35,8 +35,6 @@ function post_build_tests() {
     echo "*** Checking ${makefile}"
     make -n -C $(dirname ${makefile}) || { failed=1; echo "--- FAIL: ${makefile}"; }
   done
-  subheader "Checking config files"
-  make -C config/prow test || { failed=1; echo "--- FAIL"; }
```
Comment (coryrc) [inline test/presubmit-tests.sh]: As inappropriate as it is for this to have been in the Makefile to begin with, wouldn't an easier solution be to just run `make -C config/prow-staging test` as well?

File: config/prow/jobs/config.yaml (hunk 3 above)
Comment (chizhg) [reply config/prow/jobs/config.yaml]: I know... But for some reason, we'll have to have at least one branch to get the config generator work. I have changed the branch name to `legacy` to be less confusing.
(Config generator is not related to this PR and should have separate discussion.)

File: test/presubmit-config-check.sh (hunk 4 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: We are using `?=` in the Makefile, like https://github.com/knative/test-infra/blob/4dac123b9a3d14d2e07b4a59fbac0b0e99a09e92/config/prow_common.mk#L32.

If these env vars are defined as what we are doing here with `export XXX=xx`, the make command will actually use them. So it actually works :)

File: test/presubmit-tests.sh (hunk 6 above)
Comment (chizhg) [reply test/presubmit-tests.sh]: Staging Prow does not have Testgrid config file. I don't want to add a special logic in the Makefile, as it's already complicated enough :(

File: config/prow/Makefile (hunk 2 above)
Comment (chizhg) [reply config/prow/Makefile]: I have removed both TESTGRID_GCS and TESTGRID_CONFIG from prow_common.mk, as they are only used in production Prow.

File: test/presubmit-config-check.sh (hunk 5 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: Done, thanks.

File: config/prow_common.mk (hunk 1 above)
Comment (chizhg) [reply config/prow_common.mk]: It seems the `git` in the prow-tests image does not support `git branch --show-current`.  I'll take a further look on this tomorrow.

**Comment by chaodaiG** [comment]
This is a nice move, thanks for doing the work!

File: config/prow/Makefile (hunk 2 above)
Comment (coryrc) [reply config/prow/Makefile]: There's a different TESTGRID_GCS in prow-staging, are you sure?

File: test/presubmit-config-check.sh (hunk 4 above)
Comment (coryrc) [reply test/presubmit-config-check.sh]: I thought that only related to make variables and not environment variables. Or do all environment variables become make variables?
I don't like that the environment could make the command misbehave, but you're not making it worse. So if it works, it works.

File: config/prow/Makefile (hunk 2 above)
Comment (coryrc) [reply config/prow/Makefile]: Oh, because it has GENERATE_TESTGRID_CONFIG set to false. Okay.

File: test/presubmit-tests.sh (hunk 6 above)
Comment (coryrc) [reply test/presubmit-tests.sh]: k

File: config/prow_common.mk (hunk 1 above)
Comment (coryrc) [reply config/prow_common.mk]: Then our git at least a year old... https://stackoverflow.com/a/55088865

File: test/presubmit-config-check.sh [hunk 7]
```diff
@@ -27,19 +34,19 @@ export TESTGRID_CONFIG=$(mktemp)
 subheader "Regenerating config for production prow and testgrid"
 make -C config/prow config
 subheader "Comparing the generated config files with the existing config files"
-diff --ignore-matching-lines="^# Copyright " "${PROW_CONFIG}" "config/prow/core/config.yaml"
-diff --ignore-matching-lines="^# Copyright " "${PROW_JOB_CONFIG}" "config/prow/jobs/config.yaml"
-diff --ignore-matching-lines="^# Copyright " "${PROW_PLUGINS}" "config/prow/core/plugins.yaml"
-diff --ignore-matching-lines="^# Copyright " "${TESTGRID_CONFIG}" "config/prow/testgrid/testgrid.yaml"
+diff_config_files "${PROW_CONFIG}" "config/prow/core/config.yaml"
+diff_config_files "${PROW_JOB_CONFIG}" "config/prow/jobs/config.yaml"
+diff_config_files "${PROW_PLUGINS}" "config/prow/core/plugins.yaml"
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: I meant the set of these three should be in a function

File: config/prow/jobs/config.yaml [hunk 8]
```diff
@@ -3428,6 +3428,70 @@ presubmits:
...
+        env:
+        - name: GOOGLE_APPLICATION_CREDENTIALS
+          value: /etc/test-account/service-account.json
+        - name: E2E_CLUSTER_REGION
+          value: us-central1
+      volumes:
+      - name: test-account
+        secret:
+          secretName: test-account
+  - name: pull-knative-test-infra-check-config
```
Comment (peterfeifanchen) [inline config/prow/jobs/config.yaml]: Are these two of the same jobs? Duplicate?

Comment (peterfeifanchen) [inline config/prow/jobs/config.yaml]: pull-knative-test-infra-check-config duplicated twice here?

Comment (coryrc) [reply config/prow/jobs/config.yaml]: Duplicate. Only the image name has changed.
The config generator is currently designed to always create two jobs for every desired one, one with older image and one with latest.

Comment (peterfeifanchen) [reply config/prow/jobs/config.yaml]: Ignore this, ironically duplicated this.

Comment (coryrc) [reply config/prow/jobs/config.yaml]: See my other comment: https://github.com/knative/test-infra/pull/1766#discussion_r386577737

File: config/prow/jobs/config.yaml [hunk 9]
```diff
@@ -3428,6 +3428,70 @@ presubmits:
...
+    agent: kubernetes
+    context: pull-knative-test-infra-check-config
+    always_run: true
+    rerun_command: "/test pull-knative-test-infra-check-config"
+    trigger: "(?m)^/test (all|pull-knative-test-infra-check-config),?(\\s+|$)"
+    decorate: true
+    path_alias: knative.dev/test-infra
+    branches:
+    - "legacy"
```
Comment (peterfeifanchen) [inline config/prow/jobs/config.yaml]: I am a bit confused, if all these jobs have "legacy" as branches, do they get run then. Why are there so many jobs marked as legacy? Is this part of the generator that takes in the prow config for the legacy branches and automatically generates all this? Like I feel there may be two changes here that is not made very clear by the commit message.

1) We seem to have gotten rid of release-0.10 and replaced it with this fake "legacy" branch. 
2) There is an extra periodic job that does the config check
3) How do I know if these changes affect the production cluster vs. staging cluster?

I don't know what the effects of 1) is and I am not sure how 2) was checked previously (manually call make?). For 3) it seems we only have a single prow config file generated right, so what is the part about production vs. staging clusters?

File: test/presubmit-config-check.sh (hunk 7 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: Done.

File: config/prow/jobs/config.yaml (hunk 9 above)
Comment (chizhg) [reply config/prow/jobs/config.yaml]: 1) Either `release-0.10` or `legacy` is added for making it work for config generator, the Prow jobs themselves do not have any effect as there is only one master branch in this repo.
(This is actually a bug in config generator, but we haven't got time to fix it)

2) We don't have an extra periodic job that does the config check. Previously we are doing this check in the presubmit-build-tests, but it was not clear and missing some important checks. This PR is moving the check to a separate Prow job.

3) We have a set of config files for both production and staging Prow, see `config/prow` and `config/prow-staging`. The only configuration files they share are the cluster configs in `config/prow/cluster` (which in the future we will also make a copy in `config/prow-staging` while we implement Prow auto-update)

File: test/presubmit-config-check.sh [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  local prow_job_config="$3"
+  local prow_env="$4"
+  diff_config_files "${prow_config}" "config/${prow_env}/core/config.yaml"
+  diff_config_files "${prow_plugin_config}" "config/${prow_env}/core/plugins.yaml"
+  diff_config_files "${prow_job_config}" "config/${prow_env}/jobs/config.yaml"
+}
+
+set -e
+
+trap 'echo "Please rerun \`make -C config/prow config\`."' ERR
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: Could you add "--- FAIL: " to the beginning of the output for this and all trap?

File: test/presubmit-config-check.sh [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+header "Checking generated config for production prow and testgrid"
+export PROW_CONFIG=$(mktemp)
+export PROW_JOB_CONFIG=$(mktemp)
+export PROW_PLUGINS=$(mktemp)
+export TESTGRID_CONFIG=$(mktemp)
+subheader "Regenerating config for production prow and testgrid"
+make -C config/prow config
+subheader "Comparing the generated config files with the existing config files"
+diff_prow_config_files "${PROW_CONFIG}" "${PROW_PLUGINS}" "${PROW_JOB_CONFIG}" "prow"
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: You use the same 3 variables for both calls; you could drop all the intermediate variables and just use the globals in the function

File: config/prow_common.mk (hunk 1 above)
Comment (chizhg) [reply config/prow_common.mk]: It's already failing in presubmit-build-tests for other PRs, see https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative_test-infra/1767/pull-knative-test-infra-build-tests/1234494216499892227#1:build-log.txt%3A655. 

I added a TODO here to use `--show-current` after we update Git in the prow-tests image. Let's do it in followup PRs?

File: test/presubmit-config-check.sh (hunk 10 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: Done.

File: test/presubmit-config-check.sh (hunk 11 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: Done.

File: test/presubmit-config-check.sh [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  --job-config-path="$(realpath "config/prow/jobs/config.yaml")" \
+  --plugin-config="$(realpath "config/prow/core/plugins.yaml")"
+
+header "Validating staging Prow config files"
+bazel run @k8s//prow/cmd/checkconfig -- \
+  --config-path="$(realpath "config/prow-staging/core/config.yaml")" \
+  --job-config-path="$(realpath "config/prow-staging/jobs/config.yaml")" \
+  --plugin-config="$(realpath "config/prow-staging/core/plugins.yaml")"
+
+trap '--- FAIL: echo "Testgrid config file has errors, please check."' ERR
```
Comment (coryrc) [inline test/presubmit-config-check.sh]: This and others should be trap 'echo "--- FAIL: Testgrid config file has errors, please check."' ERR

File: config/prow_common.mk (hunk 1 above)
Comment (coryrc) [reply config/prow_common.mk]: k

File: test/presubmit-config-check.sh (hunk 12 above)
Comment (chizhg) [reply test/presubmit-config-check.sh]: My bad.. Fixed.

**Pull request merged**